bn_mp_mul.c

```diff
@@ -6,82 +6,69 @@
 /* high level multiplication (handles sign) */
 int mp_mul(const mp_int *a, const mp_int *b, mp_int *c)
 {
-   int     res, neg;
-#ifdef BN_MP_BALANCE_MUL_C
-   int len_b, len_a;
-#endif
+   int res, neg, min_len, digs;
+   min_len = MIN(a->used, b->used);
+   digs = a->used + b->used + 1;
    neg = (a->sign == b->sign) ? MP_ZPOS : MP_NEG;
-#ifdef BN_MP_BALANCE_MUL_C
-   len_a = a->used;
-   len_b = b->used;
 
-   if (len_a == len_b) {
-      goto GO_ON;
-   }
-   /*
-    * Check sizes. The smaller one needs to be larger than the Karatsuba cut-off.
-    * The bigger one needs to be at least about one KARATSUBA_MUL_CUTOFF bigger
-    * to make some sense, but it depends on architecture, OS, position of the
-    * stars... so YMMV.
-    * Using it to cut the input into slices small enough for fast_s_mp_mul_digs
-    * was actually slower on the author's machine, but YMMV.
-    */
-   if ((MIN(len_a, len_b) < KARATSUBA_MUL_CUTOFF)
-       || ((MAX(len_a, len_b)) / 2 < KARATSUBA_MUL_CUTOFF)) {
-      goto GO_ON;
-   }
-   /*
-    * Not much effect was observed below a ratio of 1:2, but again: YMMV.
-    */
-   if ((MAX(len_a, len_b) /  MIN(len_a, len_b)) < 2) {
-      goto GO_ON;
+   if (0) {
    }
 
-   res = mp_balance_mul(a,b,c);
-   goto END;
+#ifdef BN_MP_BALANCE_MUL_C
+   else if ((a->used != b->used) &&
+
+            /* Check sizes. The smaller one needs to be larger than the Karatsuba cut-off.
```

Comment:
it's about the entire readability and such complex if statements suck

Reply:
Please compare the old code with the new code side by side, don't look at the diff only.
Then I think this PR improves things.

Reply:
which old code? pre-206-merged?

Reply:
Compare
https://github.com/libtom/libtommath/blob/develop/bn_mp_mul.c
with
https://github.com/libtom/libtommath/blob/41d3d73b8e19531fd476cc57355bdb1a015c5eb3/bn_mp_mul.c
and then tell me seriously if you defend the goto version.

Reply:
Btw. I've the impression that this is going in the same direction as the rebase discussion.

Linearity of things, where linearity in my opinion leads to easier readability for a human.

Reply:
I only see superficial similarities besides you disagreeing with me on basic things. And btw goto can be much less linear while if/else etc is linear by construction.

Reply:
> besides you disagreeing with me on basic things

please explain

Reply:
>  and then tell me seriously if you defend the goto version.

BTW yes, overall it improved

Reply:
For example we disagreed about what should be tested or not, this is quite basic. Then there was the discussion about the rebasing, then the discussion about about the public ABI.

I would prefer if we wouldn't over discuss too many things to be honest. If there is something to be learned ok but otherwise it just costs time. It happened a few times now - I created a (from my pov) quick and uncontroversial patch and then a longish discussion emerges.  But I am not saying I am innocent in this respect ;)

Reply:
The length of the patch should be proportional to the length of the discussion :)

Reply:
> BTW yes, overall it improved

good to hear :sweat_smile:

Reply:
> The length of the patch should be proportional to the length of the discussion :)

True, if there are no fundamentals being discussed as e.g. the testing.

> then the discussion about about the public ABI

You mean #172 ?

> I created a (from my pov) quick and uncontroversial patch

I was mostly fine with the original patch, besides the points I commented on. Then you made a v2 which implemented `mp_mul` entirely different and also changed parts that weren't commented on... yes this leads to re-discussion of things and new discussions...